![title](https://github.com/mtinti/TriTrypML_code/blob/master/static/ANTI-INFECTIVES-RGB_200pc.png)
# TriTrypML

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/mtinti/TriTrypML_code/master)

https://zenodo.org/record/3517812

# Multi-omics integrations of dataset deposited at TriTrypDB
## Affiliation
    Michele Tinti and Michael A.J. Ferguson

    Wellcome Centre for Anti-Infectives Research
    School of Life Sciences, University of Dundee

## Abstract
<p style='text-align: justify;'>
The number of Proteomics and Genomics experiments is growing. 
This large amount of data poses new challenges to the scientific community. 
In particular, the amount of available data is not matched by the capacity of 
common bioinformatics tools for retrieving and extracting useful insights from the 
publicly accessible datasets. Machine learning can mitigate this issue by providing powerful 
analysis tools. However, data needs to be prepared and formatted for machine learning analysis. 
In this paper, we share with the trypanosome research community a series of Jupyter notebooks to 
demonstrate the utility of machine learning applied to the analysis of our own OMIC experiments 
and those deposited at the TryTripDB database.</p>

### Static versions of the notebooks are avaiable at:
https://tritrypdb-ml.onrender.com/make_dataset
https://tritrypdb-ml.onrender.com/example_1
