![title](https://github.com/mtinti/TriTrypML_code/blob/master/static/ANTI-INFECTIVES-RGB_200pc.png)
# TriTrypML

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/mtinti/TriTrypML_code/master)
[![DOI](https://zenodo.org/badge/3517812.svg)](https://zenodo.org/badge/latestdoi/3517812)

# Multi-omics integrations of dataset deposited at TriTrypDB
## Affiliation
    Michele Tinti and Michael A.J. Ferguson

    Wellcome Centre for Anti-Infectives Research
    School of Life Sciences, University of Dundee

## Abstract
<p style='text-align: justify;'>
The number of Proteomics and Genomics experiments performed in laboratories is growing exponentially 
year after year. This massive amount of data pose new challenges to the scientific community. 
In particular, the amount of available data is not matched by the capacity of bioinformatics 
tools for retrieving and extracting useful insights from the publicly accessible datasets. 
On the other Hand, Machine Learning (ML) can mitigate this issue by providing powerful analysis tools.  
However, data needs to be prepared and formatted for ML analysis. In this repository, we share with
 the Trypanosome community a series of Jupyter notebooks to demonstrate the utility of ML applied to 
 the analysis of the OMIC experiments produced in our laboratory or deposited at the TryTripDB database</p>

